src/bpf/async_event_filter.c

```diff
@@ -0,0 +1,63 @@
+#include <linux/bpf.h>
+#include <linux/bpf_perf_event.h>
+#include <bpf/bpf_helpers.h>
+#include <stdint.h>
+
+const uint32_t REGISTER_COUNT = sizeof(struct pt_regs)/sizeof(uint64_t);
+
+struct {
+    __uint(type, BPF_MAP_TYPE_ARRAY);
+    __uint(max_entries, REGISTER_COUNT);
+    __uint(map_flags, BPF_F_MMAPABLE);
+    __type(key, uint32_t);
+    __type(value, uint64_t);
+} registers SEC(".maps");
+
+struct {
+    __uint(type, BPF_MAP_TYPE_ARRAY);
+    __uint(max_entries, 1);
+    __uint(map_flags, BPF_F_MMAPABLE);
+    __type(key, uint32_t);
+    __type(value, uint64_t);
+} skips SEC(".maps");
+
+SEC("perf_event")
+int match_registers(struct bpf_perf_event_data* event) {
+#define CHECK_REG(name)                                                        \
+    {                                                                          \
+        const uint32_t i = offsetof(struct pt_regs, name) / sizeof(uint64_t);  \
+        uint64_t* reg = bpf_map_lookup_elem(&registers, &i);                   \
+        if (!reg) {                                                            \
+            return 1;                                                          \
+        }                                                                      \
+        if (event->regs.name != *reg) {                                        \
+            const uint32_t j = 0;                                              \
+            uint64_t* s = bpf_map_lookup_elem(&skips, &j);                     \
+            if (s) {                                                           \
+                *s += 1;                                                       \
+            }                                                                  \
+            return 0;                                                          \
+        }                                                                      \
+    }
+
+    CHECK_REG(r15)
```

Comment:
If we did the `do { ... } while(0)` thing so `CHECK_REG()` was followed by a semicolon then syntax highlighting would look less drunk.